pkg/lib/catalogsource/catalogsource_update.go

```diff
@@ -0,0 +1,173 @@
+package catalogsource
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// mu is a package scoped mutex for synchronizing catalog source updates
+var mu sync.Mutex
+
+/* UpdateStatus can be used to safely update the status of the provided catalog source. Note that the
+status values are updated to the values from catsrc in their entirety when using this function.
+
+• logger: used to log errors only
+
+• client: used to fetch / update catalog source status
+
+• catsrc: the CatalogSource to use as a source for status updates. Callers are
+responsible for updating the catalog source status values as necessary.
+*/
+func UpdateStatus(logger *logrus.Entry, client versioned.Interface, catsrc *v1alpha1.CatalogSource) error {
+	mu.Lock()
+	defer mu.Unlock()
+
+	// get the absolute latest update of this catalog source in case it changed
+	latest, err := client.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Get(context.TODO(), catsrc.GetName(), metav1.GetOptions{})
+	if err != nil {
+		logger.WithError(err).Error("UpdateStatus - error getting latest CatalogSource... cannot update image reference")
+		return err
+	}
+
+	// make copy (even though we're not making changes and using the status values as-is)
+	out := latest.DeepCopy()
+	out.Status = catsrc.Status
+
+	// make the status update if possible
+	if _, err := client.OperatorsV1alpha1().CatalogSources(out.GetNamespace()).UpdateStatus(context.TODO(), out, metav1.UpdateOptions{}); err != nil {
+		logger.WithError(err).Error("UpdateStatus - error while setting CatalogSource status")
+		return err
+	}
+
+	return nil
+}
+
+/* UpdateStatusCondition can be used to safely update the status conditions for the provided catalog source.
+This function will make no other changes to the status.
+
+• logger: used to log errors only
+
+• client: used to fetch / update catalog source status
+
+• catsrc: the CatalogSource to use as a source for status updates.
+
+• conditions: condition values to be updated
+*/
+func UpdateStatusCondition(logger *logrus.Entry, client versioned.Interface, catsrc *v1alpha1.CatalogSource, conditions ...metav1.Condition) error {
+	mu.Lock()
+	defer mu.Unlock()
+
+	// get the absolute latest update of this catalog source in case it changed
+	latest, err := client.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Get(context.TODO(), catsrc.GetName(), metav1.GetOptions{})
```

Comment:
I found that without the `Get` followed `UpdateStatus` there were a lot more instances of update failures in the logs, so I assumed that's why the original code (prior to my PR) did this two step approach. If you think it is better to "fail" this way, I can make the change. I assume this applies to all of the `UpdateXXX` functions I've created in this PR.

Reply:
> If you think it is better to "fail" this way, I can make the change

In the abstract, I think the controller needs to be given a chance to review any changes that occurred to the cluster since it formulated its changes. In this specific case, I think the current code will work because there are no other controllers that own the `status.conditions` field. In the long run, if we add independent controllers that touch independent conditions -- e.g. a controller deployed to a separate process -- then we can start to run into problems with data consistency; one controller sets the condition it cares about, the other inadvertently removes it.

You can probably avoid the

Reply:
OK... I'll remove the mutex as well as the necessary CatalogSource `Get` calls in all of the `UpdateXXX` and `RemoveStatusConditions` functions

Reply:
fixed